src/control/server/instance_storage_rpc.go

```diff
@@ -157,20 +162,121 @@ func (ei *EngineInstance) StorageFormatNVMe() (cResults proto.NvmeControllerResu
 	return
 }
 
-func smdGetHealth(ctx context.Context, ei *EngineInstance, dev *ctlpb.SmdDevice) error {
-	state := dev.Ctrlr.DevState
+func populateCtrlrHealth(ctx context.Context, ei *EngineInstance, req *ctlpb.BioHealthReq, ctrlr *ctlpb.NvmeController) (bool, error) {
+	state := ctrlr.DevState
 	if state == ctlpb.NvmeDevState_NEW {
-		ei.log.Debugf("skip fetching health stats on device %q in NEW state", dev, state)
-		return nil
+		ei.log.Noticef("skip fetching health stats on device %q in NEW state", ctrlr, state)
+		return false, nil
 	}
 
-	health, err := ei.GetBioHealth(ctx, &ctlpb.BioHealthReq{DevUuid: dev.Uuid})
+	health, err := getCtrlrHealth(ctx, ei, req)
 	if err != nil {
-		return errors.Wrapf(err, "device %q, state %q", dev, state)
+		return false, errors.Wrapf(err, "retrieve health stats for %q (state %q)", ctrlr, state)
+	}
+	ctrlr.HealthStats = health
+
+	return true, nil
+}
+
+// Scan SMD devices over dRPC and reconstruct NVMe scan response from results.
+func scanEngineBdevsOverDrpc(ctx context.Context, ei *EngineInstance, pbReq *ctlpb.ScanNvmeReq) (*ctlpb.ScanNvmeResp, error) {
+	scanSmdResp, err := scanSmd(ctx, ei, &ctlpb.SmdDevReq{})
+	if err != nil {
+		return nil, errors.Wrap(err, "scan smd")
+	}
+
+	// Re-link SMD devices inside NVMe controller structures and populate scan response.
+
+	pbResp := ctlpb.ScanNvmeResp{
+		State: new(ctlpb.ResponseState),
+	}
+	seenCtrlrs := make(map[string]*ctlpb.NvmeController)
+
+	for _, sd := range scanSmdResp.Devices {
+		if sd.Ctrlr == nil {
+			return nil, errors.Errorf("smd %q has no ctrlr ref", sd.Uuid)
+		}
+
+		addr := sd.Ctrlr.PciAddr
+
+		if _, exists := seenCtrlrs[addr]; !exists {
+			c := new(ctlpb.NvmeController)
+			*c = *sd.Ctrlr
+			c.SmdDevices = nil
+			c.HealthStats = nil
+			seenCtrlrs[addr] = c
+			pbResp.Ctrlrs = append(pbResp.Ctrlrs, c)
+		}
+
+		c := seenCtrlrs[addr]
+
+		// Populate health if requested.
+		healthUpdated := false
+		if pbReq.Health {
+			bhReq := &ctlpb.BioHealthReq{
+				DevUuid:  sd.Uuid,
+				MetaSize: pbReq.MetaSize,
+				RdbSize:  pbReq.RdbSize,
+			}
+			upd, err := populateCtrlrHealth(ctx, ei, bhReq, c)
+			if err != nil {
+				return nil, err
+			}
+			healthUpdated = upd
+		}
+
+		// Populate SMD (meta) if requested.
+		if pbReq.Meta {
+			nsd := new(ctlpb.SmdDevice)
+			*nsd = *sd
+			nsd.Ctrlr, nsd.MetaSize, nsd.RdbSize = nil, pbReq.MetaSize, pbReq.RdbSize
```

Comment:
I know this is legal Go, but this multiple-assignment line is sort of confusing to read. Could it be broken up into separate assignments?

Reply:
sure